From ef9b54cfe652c19ad38f19bf4fd13a9e7a9705b2 Mon Sep 17 00:00:00 2001
From: Harald Hoyer <harald@redhat.com>
Date: Sat, 30 Jun 2012 11:38:24 +0200
Subject: [PATCH] new testsuite test TEST-03-USR-MOUNT

mounts /usr as a btrfs subvolume and should respect "rw" and "ro" kernel
command line arguments properly.
---
 test/TEST-03-USR-MOUNT/99-idesymlinks.rules |    8 ++
 test/TEST-03-USR-MOUNT/Makefile             |   10 ++
 test/TEST-03-USR-MOUNT/create-root.sh       |   29 ++++++
 test/TEST-03-USR-MOUNT/cryptroot-ask.sh     |    6 ++
 test/TEST-03-USR-MOUNT/fstab                |    2 +
 test/TEST-03-USR-MOUNT/hard-off.sh          |    3 +
 test/TEST-03-USR-MOUNT/test-init.sh         |   31 +++++++
 test/TEST-03-USR-MOUNT/test.sh              |  132 +++++++++++++++++++++++++++
 8 files changed, 221 insertions(+)
 create mode 100644 test/TEST-03-USR-MOUNT/99-idesymlinks.rules
 create mode 100644 test/TEST-03-USR-MOUNT/Makefile
 create mode 100755 test/TEST-03-USR-MOUNT/create-root.sh
 create mode 100755 test/TEST-03-USR-MOUNT/cryptroot-ask.sh
 create mode 100644 test/TEST-03-USR-MOUNT/fstab
 create mode 100755 test/TEST-03-USR-MOUNT/hard-off.sh
 create mode 100755 test/TEST-03-USR-MOUNT/test-init.sh
 create mode 100755 test/TEST-03-USR-MOUNT/test.sh

diff --git a/test/TEST-03-USR-MOUNT/99-idesymlinks.rules b/test/TEST-03-USR-MOUNT/99-idesymlinks.rules
new file mode 100644
index 0000000..d557790
--- /dev/null
+++ b/test/TEST-03-USR-MOUNT/99-idesymlinks.rules
@@ -0,0 +1,8 @@
+ACTION=="add|change", SUBSYSTEM=="block", ENV{DEVTYPE}=="disk", KERNEL=="hda", SYMLINK+="sda"
+ACTION=="add|change", SUBSYSTEM=="block", ENV{DEVTYPE}=="partition", KERNEL=="hda*", SYMLINK+="sda$env{MINOR}"
+ACTION=="add|change", SUBSYSTEM=="block", ENV{DEVTYPE}=="disk", KERNEL=="hdb", SYMLINK+="sdb"
+ACTION=="add|change", SUBSYSTEM=="block", ENV{DEVTYPE}=="partition", KERNEL=="hdb*", SYMLINK+="sdb$env{MINOR}"
+ACTION=="add|change", SUBSYSTEM=="block", ENV{DEVTYPE}=="disk", KERNEL=="hdc", SYMLINK+="sdc"
+ACTION=="add|change", SUBSYSTEM=="block", ENV{DEVTYPE}=="partition", KERNEL=="hdc*", SYMLINK+="sdc$env{MINOR}"
+ACTION=="add|change", SUBSYSTEM=="block", ENV{DEVTYPE}=="disk", KERNEL=="hdd", SYMLINK+="sdd"
+ACTION=="add|change", SUBSYSTEM=="block", ENV{DEVTYPE}=="partition", KERNEL=="hdd*", SYMLINK+="sdd$env{MINOR}"
diff --git a/test/TEST-03-USR-MOUNT/Makefile b/test/TEST-03-USR-MOUNT/Makefile
new file mode 100644
index 0000000..bc0ddb6
--- /dev/null
+++ b/test/TEST-03-USR-MOUNT/Makefile
@@ -0,0 +1,10 @@
+all:
+	@make -s --no-print-directory -C ../.. all
+	@basedir=../.. testdir=../ ./test.sh --all
+setup:
+	@make --no-print-directory -C ../.. all
+	@basedir=../.. testdir=../ ./test.sh --setup
+clean:
+	@basedir=../.. testdir=../ ./test.sh --clean
+run:
+	@basedir=../.. testdir=../ ./test.sh --run
diff --git a/test/TEST-03-USR-MOUNT/create-root.sh b/test/TEST-03-USR-MOUNT/create-root.sh
new file mode 100755
index 0000000..6662bca
--- /dev/null
+++ b/test/TEST-03-USR-MOUNT/create-root.sh
@@ -0,0 +1,29 @@
+#!/bin/sh
+# don't let udev and this script step on eachother's toes
+for x in 64-lvm.rules 70-mdadm.rules 99-mount-rules; do
+    > "/etc/udev/rules.d/$x"
+done
+rm /etc/lvm/lvm.conf
+udevadm control --reload-rules
+set -e
+# save a partition at the beginning for future flagging purposes
+sfdisk -C 5120 -H 2 -S 32 -L /dev/sda <<EOF
+,16
+,
+EOF
+
+mkfs.btrfs -L dracut /dev/sda2
+btrfs device scan /dev/sda2
+mkdir -p /root
+mount -t btrfs /dev/sda2 /root
+btrfs subvolume create /root/usr
+[ -d /root/usr ] || mkdir /root/usr
+mount -t btrfs -o subvol=usr /dev/sda2 /root/usr
+cp -a -t /root /source/*
+mkdir -p /root/run
+umount /root/usr
+umount /root
+echo "dracut-root-block-created" >/dev/sda1
+sync
+poweroff -f
+
diff --git a/test/TEST-03-USR-MOUNT/cryptroot-ask.sh b/test/TEST-03-USR-MOUNT/cryptroot-ask.sh
new file mode 100755
index 0000000..db27c5b
--- /dev/null
+++ b/test/TEST-03-USR-MOUNT/cryptroot-ask.sh
@@ -0,0 +1,6 @@
+#!/bin/sh
+
+[ -b /dev/mapper/$2 ] && exit 0
+echo -n test >/keyfile
+/sbin/cryptsetup luksOpen $1 $2 </keyfile
+
diff --git a/test/TEST-03-USR-MOUNT/fstab b/test/TEST-03-USR-MOUNT/fstab
new file mode 100644
index 0000000..0e0a0e0
--- /dev/null
+++ b/test/TEST-03-USR-MOUNT/fstab
@@ -0,0 +1,2 @@
+/dev/sda2	/                       btrfs   defaults         0 0
+/dev/sda2	/usr                    btrfs   subvol=usr,ro    0 0
diff --git a/test/TEST-03-USR-MOUNT/hard-off.sh b/test/TEST-03-USR-MOUNT/hard-off.sh
new file mode 100755
index 0000000..12c3d5a
--- /dev/null
+++ b/test/TEST-03-USR-MOUNT/hard-off.sh
@@ -0,0 +1,3 @@
+#!/bin/sh
+getarg rd.shell || poweroff -f
+getarg failme && poweroff -f
diff --git a/test/TEST-03-USR-MOUNT/test-init.sh b/test/TEST-03-USR-MOUNT/test-init.sh
new file mode 100755
index 0000000..494313b
--- /dev/null
+++ b/test/TEST-03-USR-MOUNT/test-init.sh
@@ -0,0 +1,31 @@
+#!/bin/sh
+>/dev/watchdog
+export PATH=/sbin:/bin:/usr/sbin:/usr/bin
+strstr() { [ "${1#*$2*}" != "$1" ]; }
+CMDLINE=$(while read line; do echo $line;done < /proc/cmdline)
+plymouth --quit
+exec </dev/console >/dev/console 2>&1
+
+ismounted() {
+    while read a m a; do
+        [ "$m" = "$1" ] && return 0
+    done < /proc/mounts
+    return 1
+}
+
+if ismounted /usr; then
+    echo "dracut-root-block-success" >/dev/sdb
+fi
+export TERM=linux
+export PS1='initramfs-test:\w\$ '
+[ -f /etc/mtab ] || ln -sfn /proc/mounts /etc/mtab
+[ -f /etc/fstab ] || ln -sfn /proc/mounts /etc/fstab
+stty sane
+echo "made it to the rootfs!"
+if strstr "$CMDLINE" "rd.shell"; then
+	strstr "$(setsid --help)" "control" && CTTY="-c"
+	setsid $CTTY sh -i
+fi
+echo "Powering down."
+mount -n -o remount,ro /
+poweroff -f
diff --git a/test/TEST-03-USR-MOUNT/test.sh b/test/TEST-03-USR-MOUNT/test.sh
new file mode 100755
index 0000000..34a120f
--- /dev/null
+++ b/test/TEST-03-USR-MOUNT/test.sh
@@ -0,0 +1,132 @@
+#!/bin/bash
+
+TEST_DESCRIPTION="root filesystem on a btrfs filesystem with /usr subvolume"
+
+KVERSION=${KVERSION-$(uname -r)}
+
+# Uncomment this to debug failures
+#DEBUGFAIL="rd.shell rd.break"
+
+client_run() {
+    local test_name="$1"; shift
+    local client_opts="$*"
+
+    echo "CLIENT TEST START: $test_name"
+
+    dd if=/dev/zero of=$TESTDIR/result bs=1M count=1
+    $testdir/run-qemu \
+	-hda $TESTDIR/root.btrfs \
+	-hdb $TESTDIR/result \
+	-m 256M -nographic \
+	-net none -kernel /boot/vmlinuz-$KVERSION \
+	-watchdog ib700 -watchdog-action poweroff \
+	-append "root=LABEL=dracut $client_opts quiet rd.retry=3 rd.info console=ttyS0,115200n81 selinux=0 rd.debug $DEBUGFAIL" \
+	-initrd $TESTDIR/initramfs.testing
+
+    if (($? != 0)); then
+	echo "CLIENT TEST END: $test_name [FAILED - BAD EXIT]"
+        return 1
+    fi
+
+    if ! grep -m 1 -q dracut-root-block-success $TESTDIR/result; then
+	echo "CLIENT TEST END: $test_name [FAILED]"
+        return 1
+    fi
+    echo "CLIENT TEST END: $test_name [OK]"
+
+}
+
+test_run() {
+    client_run "no option specified, should fail" && return 1
+    client_run "readonly root" "ro" || return 1
+    client_run "writeable root" "rw" || return 1
+    return 0
+}
+
+test_setup() {
+    rm -f $TESTDIR/root.btrfs
+    # Create the blank file to use as a root filesystem
+    dd if=/dev/null of=$TESTDIR/root.btrfs bs=1M seek=160
+
+    kernel=$KVERSION
+    # Create what will eventually be our root filesystem onto an overlay
+    (
+	initdir=$TESTDIR/overlay/source
+	mkdir -p $initdir
+	. $basedir/dracut-functions.sh
+	dracut_install sh df free ls shutdown poweroff stty cat ps ln ip route \
+	    mount dmesg ifconfig dhclient mkdir cp ping dhclient \
+	    umount strace less setsid
+	for _terminfodir in /lib/terminfo /etc/terminfo /usr/share/terminfo; do
+            [ -f ${_terminfodir}/l/linux ] && break
+	done
+	dracut_install -o ${_terminfodir}/l/linux
+	inst "$basedir/modules.d/40network/dhclient-script.sh" "/sbin/dhclient-script"
+	inst "$basedir/modules.d/40network/ifup.sh" "/sbin/ifup"
+	dracut_install grep
+        inst_simple ./fstab /etc/fstab
+	inst ./test-init.sh /sbin/init
+	find_binary plymouth >/dev/null && dracut_install plymouth
+	(cd "$initdir"; mkdir -p dev sys proc etc var/run tmp )
+	cp -a /etc/ld.so.conf* $initdir/etc
+	sudo ldconfig -r "$initdir"
+    )
+
+    # second, install the files needed to make the root filesystem
+    (
+	initdir=$TESTDIR/overlay
+	. $basedir/dracut-functions.sh
+	dracut_install sfdisk mkfs.btrfs btrfs poweroff cp umount sync
+	inst_hook initqueue 01 ./create-root.sh
+	inst_simple ./99-idesymlinks.rules /etc/udev/rules.d/99-idesymlinks.rules
+    )
+
+    # create an initramfs that will create the target root filesystem.
+    # We do it this way so that we do not risk trashing the host mdraid
+    # devices, volume groups, encrypted partitions, etc.
+    $basedir/dracut.sh -l -i $TESTDIR/overlay / \
+	-m "dash udev-rules btrfs base rootfs-block kernel-modules" \
+	-d "piix ide-gd_mod ata_piix btrfs sd_mod" \
+        --nomdadmconf \
+        --nohardlink \
+	-f $TESTDIR/initramfs.makeroot $KVERSION || return 1
+
+    # Invoke KVM and/or QEMU to actually create the target filesystem.
+
+#    echo $TESTDIR/overlay
+#    echo $TESTDIR/initramfs.makeroot
+#exit 1
+    rm -rf $TESTDIR/overlay
+
+    $testdir/run-qemu \
+	-hda $TESTDIR/root.btrfs \
+	-m 256M -nographic -net none \
+	-kernel "/boot/vmlinuz-$kernel" \
+	-append "root=/dev/dracut/root rw rootfstype=btrfs quiet console=ttyS0,115200n81 selinux=0" \
+	-initrd $TESTDIR/initramfs.makeroot  || return 1
+    grep -m 1 -q dracut-root-block-created $TESTDIR/root.btrfs || return 1
+
+
+    (
+	initdir=$TESTDIR/overlay
+	. $basedir/dracut-functions.sh
+	dracut_install poweroff shutdown
+	inst_hook emergency 000 ./hard-off.sh
+	inst_simple ./99-idesymlinks.rules /etc/udev/rules.d/99-idesymlinks.rules
+    )
+    sudo $basedir/dracut.sh -l -i $TESTDIR/overlay / \
+	-a "debug watchdog" \
+        -o "network" \
+	-d "piix ide-gd_mod ata_piix btrfs sd_mod ib700wdt" \
+	-f $TESTDIR/initramfs.testing $KVERSION || return 1
+
+    rm -rf $TESTDIR/overlay
+
+#	-o "plymouth network md dmraid multipath fips caps crypt btrfs resume dmsquash-live dm"
+}
+
+test_cleanup() {
+    return 0
+}
+
+. $testdir/test-functions
